Mathlib/MeasureTheory/Measure/WithDensity.lean

```diff
@@ -269,51 +270,90 @@ theorem withDensity_apply_eq_zero {f : α → ℝ≥0∞} {s : Set α} (hf : Mea
     apply measure_mono_null A (measure_union_null _ _)
     · apply withDensity_absolutelyContinuous
       rwa [measure_toMeasurable]
-    · have M : MeasurableSet { x : α | f x = 0 } := hf (measurableSet_singleton _)
-      rw [withDensity_apply _ M, lintegral_eq_zero_iff hf]
-      filter_upwards [ae_restrict_mem M]
-      simp only [imp_self, Pi.zero_apply, imp_true_iff]
+    rcases hf with ⟨g, hg, hfg⟩
+    have t : {x | f x = 0} =ᵐ[μ.withDensity f] {x | g x = 0} := by
+      apply withDensity_absolutelyContinuous
+      filter_upwards [hfg] with a ha
+      rw [eq_iff_iff]
+      exact ⟨fun h ↦ by rw [h] at ha; exact ha.symm,
+             fun h ↦ by rw [h] at ha; exact ha⟩
+    rw [measure_congr t, withDensity_congr_ae hfg]
+    have M : MeasurableSet { x : α | g x = 0 } := hg (measurableSet_singleton _)
+    rw [withDensity_apply _ M, lintegral_eq_zero_iff hg]
+    filter_upwards [ae_restrict_mem M]
+    simp only [imp_self, Pi.zero_apply, imp_true_iff]
+
+theorem withDensity_apply_eq_zero {f : α → ℝ≥0∞} {s : Set α} (hf : Measurable f) :
+    μ.withDensity f s = 0 ↔ μ ({ x | f x ≠ 0 } ∩ s) = 0 :=
+  withDensity_apply_eq_zero' <| Measurable.aemeasurable hf
 #align measure_theory.with_density_apply_eq_zero MeasureTheory.withDensity_apply_eq_zero
 
-theorem ae_withDensity_iff {p : α → Prop} {f : α → ℝ≥0∞} (hf : Measurable f) :
+theorem ae_withDensity_iff' {p : α → Prop} {f : α → ℝ≥0∞} (hf : AEMeasurable f μ) :
     (∀ᵐ x ∂μ.withDensity f, p x) ↔ ∀ᵐ x ∂μ, f x ≠ 0 → p x := by
-  rw [ae_iff, ae_iff, withDensity_apply_eq_zero hf, iff_iff_eq]
+  rw [ae_iff, ae_iff, withDensity_apply_eq_zero' hf, iff_iff_eq]
   congr
   ext x
   simp only [exists_prop, mem_inter_iff, iff_self_iff, mem_setOf_eq, not_forall]
-#align measure_theory.ae_with_density_iff MeasureTheory.ae_withDensity_iff
 
-theorem ae_withDensity_iff_ae_restrict {p : α → Prop} {f : α → ℝ≥0∞} (hf : Measurable f) :
+theorem ae_withDensity_iff {p : α → Prop} {f : α → ℝ≥0∞} (hf : Measurable f) :
+    (∀ᵐ x ∂μ.withDensity f, p x) ↔ ∀ᵐ x ∂μ, f x ≠ 0 → p x :=
+  ae_withDensity_iff' <| Measurable.aemeasurable hf
+
+theorem ae_withDensity_iff_ae_restrict' {p : α → Prop} {f : α → ℝ≥0∞}
+    (hf : AEMeasurable f μ) :
     (∀ᵐ x ∂μ.withDensity f, p x) ↔ ∀ᵐ x ∂μ.restrict { x | f x ≠ 0 }, p x := by
-  rw [ae_withDensity_iff hf, ae_restrict_iff']
+  rw [ae_withDensity_iff' hf, ae_restrict_iff'₀]
   · simp only [mem_setOf]
-  · exact hf (measurableSet_singleton 0).compl
+  · rcases hf with ⟨g, hg, hfg⟩
+    have nonneg_eq_ae : {x | g x ≠ 0} =ᵐ[μ] {x | f x ≠ 0} := by {
+      filter_upwards [hfg] with a ha
+      simp only [eq_iff_iff]
+      exact ⟨fun (h : g a ≠ 0) ↦ by rwa [← ha] at h,
+             fun (h : f a ≠ 0) ↦ by rwa [ha] at h⟩
+    }
```

Comment:
```suggestion
    have nonneg_eq_ae : {x | g x ≠ 0} =ᵐ[μ] {x | f x ≠ 0} := by
      filter_upwards [hfg] with a ha
      simp only [eq_iff_iff]
      exact ⟨fun (h : g a ≠ 0) ↦ by rwa [← ha] at h,
             fun (h : f a ≠ 0) ↦ by rwa [ha] at h⟩
```

Reply:
I'm not sure to see what are the differences

Reply:
My suggestion removes `{` and `}` around the proof of the `have` statement.